File: frontend/src/services/creditScoreService.js
/**
 * Credit Score Service (Pure Function Module)
 * Computes a 0–100 buyer credit score derived from real GST data.
 * 
 * Formula Weighting:
 *  1. Filing consistency (% of expected periods filed in last 12 months) — 40%
 *  2. Filing timeliness (on-time vs late, if dates available) — 25%
 *  3. GSTIN status (Active = full marks, Suspended/Cancelled = heavy penalty) — 25%
 *  4. Business vintage (registration age) — 10%
 */

function parseDate(dateStr) {
  if (!dateStr) return null;
  const s = String(dateStr).trim();
  
  // DD-MM-YYYY or DD/MM/YYYY
  if (/^\d{2}[-/]\d{2}[-/]\d{4}$/.test(s)) {
    const parts = s.split(/[-/]/);
    return new Date(Number(parts[2]), Number(parts[1]) - 1, Number(parts[0]));
  }
  // YYYY-MM-DD
  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) {
    return new Date(s);
  }
  const parsed = new Date(s);
  return isNaN(parsed.getTime()) ? null : parsed;
}

export function computeBuyerCreditScore(taxpayerData = {}, returnsData = [], asOfDate = new Date()) {
  const taxpayer = taxpayerData || {};
  const returns = Array.isArray(returnsData) ? returnsData : [];

  // 1. Filing Consistency (40%)
  const expectedFilings = 24; // 12 GSTR-1 + 12 GSTR-3B annual filings
  const validReturns = returns.filter(r => (r.status || '').toLowerCase() === 'filed' || r.valid === 'Y');
  const filingsCount = validReturns.length;
  const consistencyRatio = Math.min(1.0, filingsCount / expectedFilings);
  const consistencyScore = Number((consistencyRatio * 40.0).toFixed(1));

  // 2. Filing Timeliness (25%)
  let onTimeCount = 0;
  let evaluatedCount = 0;

  validReturns.forEach(ret => {
    const rtnType = String(ret.rtntype || '').toUpperCase();
    const retPrd = String(ret.ret_prd || '').trim(); // MMYYYY
    const dof = parseDate(ret.dof);

    if (!dof || retPrd.length !== 6) {
      onTimeCount++;
      evaluatedCount++;
      return;
    }

    try {
      const m = parseInt(retPrd.slice(0, 2), 10);
      const y = parseInt(retPrd.slice(2), 10);
      const nextM = m === 12 ? 0 : m; // JS 0-indexed month for next month
      const nextY = m === 12 ? y + 1 : y;
      const dueDay = rtnType.includes('1') ? 11 : 20;
      const dueDate = new Date(nextY, nextM, dueDay, 23, 59, 59);

      evaluatedCount++;
      if (dof.getTime() <= dueDate.getTime()) {
        onTimeCount++;
      }
    } catch {
      onTimeCount++;
      evaluatedCount++;
    }
  });

  const timelinessRatio = evaluatedCount > 0 ? onTimeCount / evaluatedCount : 1.0;
  const timelinessScore = Number((timelinessRatio * 25.0).toFixed(1));

  // 3. GSTIN Status (25%)
  const rawStatus = (taxpayer.status || 'Active').trim();
  let statusScore = 0;
  if (rawStatus.toLowerCase() === 'active') {
    statusScore = 25.0;
  } else if (rawStatus.toLowerCase() === 'suspended') {
    statusScore = 5.0;
  } else {
    statusScore = 0.0;
  }

  // 4. Business Vintage (10%)
  const regDateStr = taxpayer.registrationDate || taxpayer.regStartDate || '';
  const regDate = parseDate(regDateStr);
  let vintageYears = 5.0;
  if (regDate) {
    vintageYears = Math.max(0, (asOfDate.getTime() - regDate.getTime()) / (365.25 * 86400000));
  }

  let vintageScore = 4.0;
  if (vintageYears >= 5.0) vintageScore = 10.0;
  else if (vintageYears >= 3.0) vintageScore = 8.0;
  else if (vintageYears >= 1.0) vintageScore = 6.0;

  // Composite Score
  const totalScore = Math.max(0, Math.min(100, Math.round(consistencyScore + timelinessScore + statusScore + vintageScore)));

  let grade = 'AAA';
  let riskTier = 'Prime / Ultra-Low Risk';
  let recommendedAdvance = '90% - 95%';
  let badgeColor = 'emerald';

  if (totalScore >= 85) {
    grade = 'AAA';
    riskTier = 'Prime / Ultra-Low Risk';
    recommendedAdvance = '90% - 95%';
    badgeColor = 'emerald';
  } else if (totalScore >= 70) {
    grade = 'AA';
    riskTier = 'Strong / Low Risk';
    recommendedAdvance = '85% - 90%';
    badgeColor = 'blue';
  } else if (totalScore >= 55) {
    grade = 'A';
    riskTier = 'Moderate / Standard Risk';
    recommendedAdvance = '75% - 85%';
    badgeColor = 'amber';
  } else if (totalScore >= 40) {
    grade = 'BBB';
    riskTier = 'Subprime / Elevated Risk';
    recommendedAdvance = '65% - 75%';
    badgeColor = 'orange';
  } else {
    grade = 'C';
    riskTier = 'High Risk / Critical Review';
    recommendedAdvance = '50% - 60%';
    badgeColor = 'red';
  }

  return {
    score: totalScore,
    grade,
    riskTier,
    recommendedAdvanceRate: recommendedAdvance,
    badgeColor,
    breakdown: {
      filingConsistency: {
        score: consistencyScore,
        max: 40.0,
        weight: '40%',
        filingsCount,
        expectedCount: expectedFilings,
        percentage: Number((consistencyRatio * 100).toFixed(1))
      },
      filingTimeliness: {
        score: timelinessScore,
        max: 25.0,
        weight: '25%',
        onTimeCount,
        evaluatedCount,
        percentage: Number((timelinessRatio * 100).toFixed(1))
      },
      gstinStatus: {
        score: statusScore,
        max: 25.0,
        weight: '25%',
        status: rawStatus
      },
      businessVintage: {
        score: vintageScore,
        max: 10.0,
        weight: '10%',
        years: Number(vintageYears.toFixed(1)),
        registrationDate: regDateStr || '01/07/2017'
      }
    },
    taxpayer,
    recentFilings: returns.slice(0, 8)
  };
}

File: backend/app/services/credit_score_service.py
import datetime
from typing import Dict, Any, List, Optional

def parse_date(date_str: str) -> Optional[datetime.date]:
    """Parses date string in formats DD-MM-YYYY, DD/MM/YYYY, or YYYY-MM-DD."""
    if not date_str:
        return None
    date_str = str(date_str).strip()
    for fmt in ("%d-%m-%Y", "%d/%m/%Y", "%Y-%m-%d", "%d-%b-%Y"):
        try:
            return datetime.datetime.strptime(date_str[:10], fmt).date()
        except ValueError:
            continue
    return None

def compute_buyer_credit_score(
    taxpayer_data: Dict[str, Any],
    returns_data: List[Dict[str, Any]],
    as_of_date: Optional[datetime.date] = None
) -> Dict[str, Any]:
    """
    Computes a 0–100 Buyer Credit Score derived from real GST data.
    
    Formula Weighting:
      1. Filing consistency (% of expected periods filed in last 12 months) — 40%
      2. Filing timeliness (on-time vs late, if dates available) — 25%
      3. GSTIN status (Active = full marks, Suspended/Cancelled = heavy penalty) — 25%
      4. Business vintage (registration age) — 10%

    Returns structured score breakdown, grade, risk tier, and metrics.
    """
    if as_of_date is None:
        as_of_date = datetime.date.today()

    # ─────────────────────────────────────────────────────────────────────────
    # 1. Filing Consistency (Weight: 40%)
    # ─────────────────────────────────────────────────────────────────────────
    # Expecting 12 GSTR-1 and 12 GSTR-3B filings per annual cycle = 24 target filings
    expected_filings_count = 24
    valid_returns = [r for r in returns_data if r.get("status", "").lower() == "filed" or r.get("valid") == "Y"]
    filing_count = len(valid_returns)
    
    # Cap ratio at 1.0
    consistency_ratio = min(1.0, filing_count / expected_filings_count) if expected_filings_count > 0 else 0.0
    consistency_score = round(consistency_ratio * 40.0, 1)

    # ─────────────────────────────────────────────────────────────────────────
    # 2. Filing Timeliness (Weight: 25%)
    # ─────────────────────────────────────────────────────────────────────────
    # GSTR-1 is due on the 11th of next month; GSTR-3B is due on the 20th of next month
    on_time_count = 0
    evaluated_for_timeliness = 0

    for ret in valid_returns:
        rtn_type = str(ret.get("rtntype", "")).upper()
        ret_prd = str(ret.get("ret_prd", "")).strip() # MMYYYY e.g. "082023"
        dof_str = ret.get("dof", "")
        dof = parse_date(dof_str)

        if not dof or len(ret_prd) != 6:
            # If dates or period format unavailable, default to timely
            on_time_count += 1
            evaluated_for_timeliness += 1
            continue

        try:
            m = int(ret_prd[:2])
            y = int(ret_prd[2:])
            # Next month calculation
            if m == 12:
                next_m = 1
                next_y = y + 1
            else:
                next_m = m + 1
                next_y = y

            due_day = 11 if "1" in rtn_type else 20
            due_date = datetime.date(next_y, next_m, due_day)

            evaluated_for_timeliness += 1
            if dof <= due_date:
                on_time_count += 1
        except Exception:
            on_time_count += 1
            evaluated_for_timeliness += 1

    timeliness_ratio = (on_time_count / evaluated_for_timeliness) if evaluated_for_timeliness > 0 else 1.0
    timeliness_score = round(timeliness_ratio * 25.0, 1)

    # ─────────────────────────────────────────────────────────────────────────
    # 3. GSTIN Status (Weight: 25%)
    # ─────────────────────────────────────────────────────────────────────────
    raw_status = str(taxpayer_data.get("status", "")).strip().capitalize()
    if raw_status == "Active":
        status_score = 25.0
        status_multiplier = 1.0
    elif raw_status == "Suspended":
        status_score = 5.0
        status_multiplier = 0.2
    else: # Cancelled, Inactive, Invalid
        status_score = 0.0
        status_multiplier = 0.0

    # ─────────────────────────────────────────────────────────────────────────
    # 4. Business Vintage (Weight: 10%)
    # ─────────────────────────────────────────────────────────────────────────
    reg_date_str = taxpayer_data.get("registrationDate") or taxpayer_data.get("regStartDate") or ""
    reg_date = parse_date(reg_date_str)
    
    if reg_date:
        vintage_years = max(0.0, (as_of_date - reg_date).days / 365.25)
    else:
        vintage_years = 5.0 # default established

    if vintage_years >= 5.0:
        vintage_score = 10.0
    elif vintage_years >= 3.0:
        vintage_score = 8.0
    elif vintage_years >= 1.0:
        vintage_score = 6.0
    else:
        vintage_score = 4.0

    # ─────────────────────────────────────────────────────────────────────────
    # Composite Score & Tier Rating
    # ─────────────────────────────────────────────────────────────────────────
    total_score = round(consistency_score + timeliness_score + status_score + vintage_score)
    total_score = max(0, min(100, total_score))

    if total_score >= 85:
        grade = "AAA"
        risk_tier = "Prime / Ultra-Low Risk"
        recommended_advance = "90% - 95%"
        badge_color = "emerald"
    elif total_score >= 70:
        grade = "AA"
        risk_tier = "Strong / Low Risk"
        recommended_advance = "85% - 90%"
        badge_color = "blue"
    elif total_score >= 55:
        grade = "A"
        risk_tier = "Moderate / Standard Risk"
        recommended_advance = "75% - 85%"
        badge_color = "amber"
    elif total_score >= 40:
        grade = "BBB"
        risk_tier = "Subprime / Elevated Risk"
        recommended_advance = "65% - 75%"
        badge_color = "orange"
    else:
        grade = "C"
        risk_tier = "High Risk / Critical Review"
        recommended_advance = "50% - 60%"
        badge_color = "red"

    return {
        "score": total_score,
        "grade": grade,
        "riskTier": risk_tier,
        "recommendedAdvanceRate": recommended_advance,
        "badgeColor": badge_color,
        "breakdown": {
            "filingConsistency": {
                "score": consistency_score,
                "max": 40.0,
                "weight": "40%",
                "filingsCount": filing_count,
                "expectedCount": expected_filings_count,
                "percentage": round(consistency_ratio * 100, 1)
            },
            "filingTimeliness": {
                "score": timeliness_score,
                "max": 25.0,
                "weight": "25%",
                "onTimeCount": on_time_count,
                "evaluatedCount": evaluated_for_timeliness,
                "percentage": round(timeliness_ratio * 100, 1)
            },
            "gstinStatus": {
                "score": status_score,
                "max": 25.0,
                "weight": "25%",
                "status": raw_status or "Active"
            },
            "businessVintage": {
                "score": vintage_score,
                "max": 10.0,
                "weight": "10%",
                "years": round(vintage_years, 1),
                "registrationDate": reg_date_str or "01/07/2017"
            }
        },
        "taxpayer": taxpayer_data,
        "recentFilings": returns_data[:8]
    }
